src/OmniSharp.Host/MSBuild/Discovery/Extensions.cs

```diff
@@ -1,52 +1,104 @@
+                    );
                 }
 
-                msbuildLocator.RegisterInstance(instanceToRegister);
+                msbuildLocator.RegisterInstance(bestInstanceFound);
             }
             else
             {
                 logger.LogError("Could not locate MSBuild instance to register with OmniSharp");
             }
         }
 
+        public static bool HasDotNetSdksResolvers(this MSBuildInstance instance)
+        {
+            const string dotnetSdkResolver = "Microsoft.DotNet.MSBuildSdkResolver";
+
+            return File.Exists(
+                Path.Combine(
+                    instance.MSBuildPath
+                    , "SdkResolvers"
+                    , dotnetSdkResolver
+                    , dotnetSdkResolver + ".dll"
+                )
+            );
+        }
 
         public static bool IsInvalidVisualStudio(this MSBuildInstance instance)
+
             // MSBuild from Visual Studio 2017 RTM cannot be used.
             => instance.Version.Major == 15
-            && instance.Version.Minor == 0
-            && (instance.DiscoveryType == DiscoveryType.DeveloperConsole
-                || instance.DiscoveryType == DiscoveryType.VisualStudioSetup);
+                && instance.Version.Minor == 0
+                && (instance.DiscoveryType == DiscoveryType.DeveloperConsole
+                    || instance.DiscoveryType == DiscoveryType.VisualStudioSetup);
+
+        public static MSBuildInstance GetBestInstance(this IMSBuildLocator msbuildLocator, out bool invalidVSFound)
+        {
+            invalidVSFound = false;
+            MSBuildInstance bestMatchInstance = null;
+            var bestMatchScore = -1;
+
+            // same as (without so many allocations)
+            // var bestMatchInstance =  msbuildLocator
+            //     .GetInstances()
+            //     .Select(i => (instance: i, featureScore: GetFeatureScore(i)))
+            //     .OrderByDescending(i => i.featureScore)
+            //     .ThenByDescending(i => i.instance.Version.Major)
+            //     .Select(i => i.instance)
+            //     .FirstOrDefault();
+
+            foreach (var instance in msbuildLocator.GetInstances())
+            {
+                var score = GetInstanceFeatureScore(instance);
+
+                invalidVSFound = invalidVSFound || instance.IsInvalidVisualStudio();
+
+                if (score > bestMatchScore
+                    || (score == bestMatchScore && instance.Version.Major > (bestMatchInstance?.Version.Major ?? 0)))
+                {
+                    bestMatchInstance = instance;
+                    bestMatchScore = score;
+                }
+            }
+
+            return bestMatchInstance;
+        }
+
+        private static int GetInstanceFeatureScore(MSBuildInstance i)
+        {
+            var score = 0;
+
+            if (i.HasDotNetSdksResolvers())
+                score++;
+
+            if (i.IsInvalidVisualStudio())
```

Comment:
It it's invalid, might as well return -1 or so, right? We don't ever want to pick this one

Reply:
Yeah, it will always be the one with the lowest score

Reply:
Done.

Reply:
I was thinking that you could put this case first and return some value that you believe to be the lowest value, making it clear that invalid versions get the worse score.